### PR TITLE
Unit testing

### PR DESCRIPTION
**Pull request by AndrewClugston**
These unit tests cover the functionality that has currently been added and some of the future development still to be implemented.  

As more functionality is added new unit tests will be added to cover this.

These unit tests currently work by running start_unit_testing.sh 

This closes issue #3

### CODE REVIEW

File: tests/unitTests/DatawatchTests.csv [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+comment,0,0,q,.m.rbd"",,,Abdul's queries here
+
+comment,0,0,q,.my.rdb"select ravgprice:avg avgs price by sym,time from trade where time within ($st;$et), sym in (`{Sym:`})",,,rdb query
+
+comment,0,0,q,.my.hdb"select high:max price, low:min price, open:first price, close:last price by sym, date from trade where date within (2018.04.04;2018.04.09), sym in (`{Sym:`}) ",,,
+comment,0,0,q,.my.rdb"select lastTrade:last price,open:first price,high:max price,low:min price,volume:sum size by sym from trade where sym in (`{Sym:`})",,,
+comment,0,0,q,.my.rdb"select by sym from trade where sym in (`{Sym:`})",,,
+comment,0,0,q,.my.rdb"select by sym from quote where sym in (`{Sym:`})",,,
+comment,0,0,q,.my.rdb"select distinct sym from trade",,,
```
Comment (AndrewClugston) [inline tests/unitTests/DatawatchTests.csv]: The types of tests run in the DatawatchTests.csv are still to be determined and are currently set as comments.

**Comment by ThomasSmyth** [comment]
Edit the description with the issue being closed.

File: start_unit_testing.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+. ./setenv.sh
+
+export KDBTESTS=${TORQHOME}/tests
+export KDBSTACKID="-stackid ${KDBBASEPORT}"
+
+#Execute Unit tests
+echo 'Executing Unit Tests...'
+q torq.q -procname unittests1 -proctype unittests -test tests/unitTests -debug
```
Comment (ThomasSmyth) [inline start_unit_testing.sh]: Will this drop you in a q session? Is it possible to get this to run, spit out a summary then exit?
(I don't know + not sure if this is out of scope).

File: tests/setHandles.q [hunk 3]
```diff
@@ -0,0 +1,7 @@
+base:"I"$getenv[`KDBBASEPORT]									//Get base port used by all processes
+
+.my.pnl:hopen `$":localhost:",string[base+2],":rdb:pass"					//Open PnL port
+.my.rtas:hopen `$":localhost:",string[base+2],":rdb:pass"					//Open Real-Time Aggregation Subscriber port
+.my.rdb:hopen `$":localhost:",string[base+2],":rdb:pass"					//Open connection to RDB
+.my.hdb:hopen `$":localhost:",string[base+2],":rdb:pass"					//Open connection to HDB
+.my.tst:hopen `$":localhost:",string[base],":rdb:pass"						//Test subscriber
```
Comment (ThomasSmyth) [inline tests/setHandles.q]: Should avoid hardcoding these + they need semicolons.

File: tests/unitTests/PnLTests.csv [hunk 4]
```diff
@@ -0,0 +1,28 @@
+comment,0,0,q,,,,Unit tests covering real-time PnL calculator
+
+beforeany,0,0,q,.my.pnl"1+1",,,Check Connection
+
+before,0,0,q,.my.pnl"`clienttrade upsert (2018.04.10D00:00:01.672348000;`TEST;2.45;129i;0b;\"P\";\"N\";`BUY)",,,Insert fake data for testing
+before,0,0,q,.my.pnl"`clienttrade upsert (2018.04.10D00:00:05.072320000;`TEST;2.39;129i;0b;\"P\";\"N\";`SELL)",,,Insert fake data for testing -7.74
+before,0,0,q,.my.pnl"`srcquote upsert (2018.04.10D00:00:07.072320000;`TEST;`TEST;22f;70f;100j;300j;\"P\";\"N\")",,,Insert fake data to check filter
+
+true,0,0,q,.my.pnl"2 = count tables[]",,,Check only srcquote and clienttrade tables exist
```
Comment (ThomasSmyth) [inline tests/unitTests/PnLTests.csv]: 2 tables may exist, but what if they aren't the 2 you're expecting?

File: tests/unitTests/PnLTests.csv [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+true,0,0,q,.my.pnl"1 < count srcquote",,,Check srcquote table has data
+true0,0,q,.my.pnl"1 < count clienttrade",,,Check clienttrade table has data
```
Comment (ThomasSmyth) [inline tests/unitTests/PnLTests.csv]: Should the check be `0<` to check if any data is present?

File: tests/unitTests/PnLTests.csv [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+
+fail,0,0,q,.my.pnl"0 > count select from srcquote where time < .z.d - 1",,,Check srcquote table has data from previous day
+fail,0,0,q,.my.pnl"0 > count select from clienttrade where time < .z.d - 1",,,Check clienttrade table has data from previous day
+fail,0,0,q,.my.pnl"select last msum price by sym from clienttrade = select totalprice from pnl",,,Check threshold for syms
+
+run,0,0,q,.my.pnl"select from pnl where time > .z.t - 00:10",,,Check for Latest PnL update
+run,0,0,q,.my.pnl"get `:pnl",,,Check data is persisted to disk
+run,0,0,q,.my.pnl"select from srcquote where i=(select id from pnl)",,,Track record quote/trade
+run,0,0,q,.my.pnl"select pnl[`total] from pnl",,,Check total PnL
```
Comment (ThomasSmyth) [inline tests/unitTests/PnLTests.csv]: Looks a bit odd, why is `` `total`` passed to `pnl`, what is its structure?

**Comment by ThomasSmyth** [comment]
Get an avatar please.

File: tests/setHandles.q (hunk 3 above)
Comment (jonathonmcmurray) [reply tests/setHandles.q]: why aren't you making use of discovery process?

File: start_unit_testing.sh (hunk 2 above)
Comment (jonathonmcmurray) [reply start_unit_testing.sh]: should be doable, I think there's a `-noredirect` flag or something like that which should be used instead of `-debug` I think
